### PR TITLE
Fixes #25948: Score breakdown has inconsistent numbers

### PR DESCRIPTION
**Pull request by VinceMacBuche**
https://issues.rudder.io/issues/25948

### CODE REVIEW

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/score/ScoreService.scala [hunk 1]
```diff
@@ -48,32 +49,33 @@ trait ScoreService {
...
   private val availableScore: Ref[List[(String, String)]] =
     Ref.make((ComplianceScore.scoreId, "Compliance") :: (SystemUpdateScore.scoreId, "System updates") :: Nil).runNow
 
-  def init(): IOResult[Unit] = {
+  def init()(implicit qc : QueryContext): IOResult[Unit] = {
     for {
+      nodeIds <- nodeFactRepo.getAll().map(_.keySet)
       globalScores <- globalScoreRepository.getAll()
-      _            <- cache.set(globalScores)
+      _            <- cache.set(globalScores.filter(n => nodeIds.contains(n._1)))
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/score/ScoreService.scala:72]: here, `nodeIds` only contains node as seen by the person issuing the query context. I think we want all (active) nodes at that point, so directly using the system query context here, and not as a parameter (even if it's called with that, I don't think it should be a parameter at all).

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/score/ScoreService.scala [hunk 2]
```diff
@@ -136,19 +138,21 @@ class ScoreServiceImpl(globalScoreRepository: GlobalScoreRepository, scoreReposi
...
+  def cleanScore(name: String)(implicit qc: QueryContext): IOResult[Unit] = {
     for {
       _         <- cache.update(_.map { case (id, gscore) => (id, gscore.copy(details = gscore.details.filterNot(_.scoreId == name))) })
       newScores <- scoreCache.updateAndGet(_.map { case (id, scores) => (id, scores.filterNot(_.scoreId == name)) })
       _         <- update(newScores)
     } yield {}
   }
 
-  def update(newScores: Map[NodeId, List[Score]]): IOResult[Unit] = {
+  def update(newScores: Map[NodeId, List[Score]])(implicit qc : QueryContext): IOResult[Unit] = {
```
Comment (fanf) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/score/ScoreService.scala:149]: same for the queryContext, I don't think we want that to dependant of a tenant view, else the cache will become inconsistant

**Comment by fanf** [comment]
Nice catch. Can you add a unit test to capture the two invariant ? (after init, repo only contains existing nodes and service only save existing nodes)?

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/score/ScoreService.scala (hunk 1 above)
Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/score/ScoreService.scala:72]: I don't know it gives access to score of all Nodes even those you cannot, like a leak of data to another tenant

Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/score/ScoreService.scala:72]: Sorry i haven't seen it was in init, it should be systemQC always (and it is always used with it)

**Comment by VinceMacBuche** [comment]
Still needs tests

**Comment by VinceMacBuche** [comment]
Commit modified

**Comment by VinceMacBuche** [comment]
Commit modified

**Comment by VinceMacBuche** [comment]
Commit modified

**Review by fanf — APPROVED** [review]
:rocket: seems to work as expected :+1:

**Comment by VinceMacBuche** [comment]
OK, merging this PR

**Pull request merged**